### PR TITLE
Autocompletion: Register depended parsers with the main parser

### LINKED ISSUE #92527: Autocompletion fails when script is attached in the current scene
### Tested versions

v4.3.beta.custom_build [8bf8f41fc]

### System information

Windows 10.0.19045 - Vulkan (Forward+) - dedicated NVIDIA GeForce GTX 1060 (NVIDIA; 31.0.15.4633) - Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz (8 Threads)

### Issue description

https://github.com/godotengine/godot/assets/2223172/b97d1cae-f6b1-4808-ac53-8c10192d0888

### Steps to reproduce

1.
```GDScript
extends Node

func _ready() -> void:
	var dup := duplicate()
	
```
2. Make a new scene and attach this script
3. Try autocompleting `dup` methods while the scene is opened

### Minimal reproduction project (MRP)

N/A

### LINKED ISSUE #94366: [4.3-beta3]: Crash in certain autocompletion situations
### Tested versions

- Reproducible in 4.3-beta3 and current master
- I was not able to reproduce it in 4.2-stable
- I tried bisecting it but was not able to reach a final conclusion
- it has been there since at least 78680cdebc8a97bc2d81decc72cd8cccafe1f669 (not related in anyway, just what I ended on at my last failed bisect attempt) but it gets increasingly harder to replicate before the changes about GDScript Cache invalidation (more on that later)

### System information

Fedora 40

### Issue description

Since I started testing 4.3 I felt like code editing got more unstable (I experienced some crashes now and then). Now I finally found a somewhat reliable way to reproduce such a crash. (I can't guarantee that this was the reason behind all the crashes I experienced, but after looking into it a bit longer, I'd say the underlying problem might make auto completion crash in quite a lot of cases.)

The specific reproducible issue:
The following code (see MRP for full script)
```gdscript
const ONE_CONSTANT_TO_RULE_THEM_ALL = "krimpatul"

func test():
    var dummy = get_script()
                  # cursor at the indented start of this line and ignore this comment
```
Can trigger the crash when typing "d" then rapidly using enter to autocomplete "dummy" and then rapidly typing "."
- Set the GDScript Autocompletion delay to minimal for this to work

This order of actions _might_ trigger a signal 11 core dumped from time to time (it's really inconsistent you will have to try often, and be quick)

The exact point of failure is this:
https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/modules/gdscript/gdscript_editor.cpp#L1087
`member.constant` apparently has no identifier

Here is a video of achieving the crash for reference:

https://github.com/user-attachments/assets/cfd1fbef-56ce-46e5-89fc-aa49e3173c75

> The mrp contains a minimal and an original script. I was able to reproduce it in `minimal.gd` but it seems to be easier to reproduce in the more complex `original.gd`.

While I have no clue why those exact conditions trigger the crash I have somewhat figured out what is going on:
- when typing "." after dummy the completion code tries to guess the type of `dummy`
- since it has no static type it tries to guess the type of the last assigned expression
- to do that it copies the completion context and changes the current line https://github.com/godotengine/godot/blob/97b8ad1af0f2b4a216f6f1263bef4fbc69e56c7b/modules/gdscript/gdscript_editor.cpp#L2241-L2244
- through a trying the method call the autocompletion gets the return value of `get_script()`
- when trying to guess the type from the value it tries to load the attached GDScript through the GDScriptCache which does not have it cached and creates a new parser
- the class type from this parser is used as type for `dummy` and the new parser gets added to the completion context dependent parsers
- but this completion context is not the original one but a copied one with a copied list of dependent parsers
- when leaving the scope this refcounted copied context gets freed, with it the list and the parser. The parser also frees the ClassNode while being destructed (This is why it is harder to reproduce with GDScript Cache invalidation. If the cache still has a cached parser for the script, it is referenced somewhere else and won't get freed when the completion context is destructed)
- now the class type of dummy is already freed when the completion tries to list out the members, in this case a constant.
- Godot Crashes

I have no clue why those specific conditions trigger it, but the thing is, that we copy the completion context often and also add dependent parsers to it in multiple code paths, so I suspect that a lot of my crashes will come back to the base problem: a copied completion context invalidates all class types that were additionally parsed, when it is destructed

### Steps to reproduce

- open the mrp
- open either `original.tscn` or `minimal.tscn` and the corresponding script (original seems to be easier to reproduce)
- set the code completion delay as low as possible
- place the cursor bellow `var dummy`
- do the following as quick as possible:
    - press "d"
    - press "enter"
    - press "."
- repeat until it crashes (see my video for reference)

- (I also save from time to time with nothing on the line below `var dummy` and wait for the idle parse delay. Not sure if it is necessary though)

### Minimal reproduction project (MRP)

[heisenbug.zip](https://github.com/user-attachments/files/16228862/heisenbug.zip)

### PR DESCRIPTION
**Pull request by HolonProduction**
Fixes #94366 (I hope :crossed_fingers: have tried for a while and wasn't able to reproduce the crash anymore.)

Seems also to fix #92527 which seems very similar to #94366 with the crash being the difference. I suspect that this issue might also trigger a crash in rare occasions but haven't tested it.
Fixes #92527

### CODE REVIEW

**Review by akien-mga — APPROVED** [review]
Seems reasonable to me.

File: modules/gdscript/gdscript_editor.cpp [hunk 1]
```diff
@@ -2310,18 +2307,18 @@ static bool _guess_identifier_type(GDScriptParser::CompletionContext &p_context,
...
+				if (parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
+					r_type.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
+					r_type.type.script_path = script;
+					r_type.type.class_type = parser->get_parser()->get_tree();
+					r_type.type.is_meta_type = true;
+					r_type.type.is_constant = false;
+					r_type.type.kind = GDScriptParser::DataType::CLASS;
+					r_type.value = Variant();
+					return true;
+				}
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:2321]:
```suggestion
			if (parser.is_valid() && parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
				r_type.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
				r_type.type.script_path = script;
				r_type.type.class_type = parser->get_parser()->get_tree();
				r_type.type.is_meta_type = true;
				r_type.type.is_constant = false;
				r_type.type.kind = GDScriptParser::DataType::CLASS;
				r_type.value = Variant();
				return true;
```
Cleaner IMO

File: modules/gdscript/gdscript_editor.cpp [hunk 2]
```diff
@@ -1849,17 +1846,17 @@ static bool _guess_expression_type(GDScriptParser::CompletionContext &p_context,
...
+															if (parser.is_valid()) {
+																if (parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
+																	r_type.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
+																	r_type.type.script_path = script;
+																	r_type.type.class_type = parser->get_parser()->get_tree();
+																	r_type.type.is_constant = false;
+																	r_type.type.kind = GDScriptParser::DataType::CLASS;
+																	r_type.value = Variant();
+																	found = true;
+																}
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1859]:
```suggestion
															if (parser.is_valid() && parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
																r_type.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
																r_type.type.script_path = script;
																r_type.type.class_type = parser->get_parser()->get_tree();
																r_type.type.is_constant = false;
																r_type.type.kind = GDScriptParser::DataType::CLASS;
																r_type.value = Variant();
																found = true;
```

File: modules/gdscript/gdscript_editor.cpp [hunk 3]
```diff
@@ -1521,22 +1521,21 @@ static GDScriptCompletionIdentifier _type_from_variant(const Variant &p_value, G
...
-					p_context.dependent_parsers.push_back(parser);
-					return ci;
+				Ref<GDScriptParserRef> parser = p_context.parser->get_depended_parser_for(scr->get_path());
+				if (parser.is_valid()) {
+					if (parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
+						ci.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
+						ci.type.class_type = parser->get_parser()->get_tree();
+						ci.type.kind = GDScriptParser::DataType::CLASS;
+						return ci;
+					}
```
Comment (AThousandShips) [inline modules/gdscript/gdscript_editor.cpp:1536]:
```suggestion
				if (parser.is_valid() && parser->raise_status(GDScriptParserRef::INTERFACE_SOLVED) == OK) {
					ci.type.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;
					ci.type.class_type = parser->get_parser()->get_tree();
					ci.type.kind = GDScriptParser::DataType::CLASS;
					return ci;
```

**Review by AThousandShips — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**